### PR TITLE
Handle FocusZone edge cases

### PR DESCRIPTION
**Pull request by chiuam**
### Platforms Impacted
- [ ] iOS
- [X] macOS
- [ ] win32 (Office)
- [ ] windows
- [ ] android

### Description of changes

Currently, keyboard navigation breaks when these two cases happen:
1. FocusZone is empty
2. FocusZone has no focusable elements

That's because `becomeFirstReesponder` is false when a nil key view is being passed in which causes the key view loop to get reset.

Let's handle said edge cases by rejecting firstResponder if FocusZone is empty or has no focusable elements.

### Verification
Added new test cases and traversed through (forward & backward directions) them in the FocusZone test page.

### Pull request checklist

This PR has considered (when applicable):
- [ ] Automated Tests
- [X] Documentation and examples
- [X] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

### CODE REVIEW

File: packages/components/FocusZone/macos/RCTFocusZone.m [hunk 1]
```diff
@@ -157,10 +157,32 @@ static inline BOOL IsHorizontalNavigationWithinZoneAction(FocusZoneAction action
...
+            return YES;
+        }
+    }
+
+    return NO;
+}
+
 - (BOOL)acceptsFirstResponder
 {
 	// Accept firstResponder on FocusZone itself in order to reassign it within the FocusZone.
```
Comment (harrieshin) [inline packages/components/FocusZone/macos/RCTFocusZone.m:184]: it seems like the comment should be updated?

File: packages/components/FocusZone/macos/RCTFocusZone.m [hunk 2]
```diff
@@ -157,10 +157,32 @@ static inline BOOL IsHorizontalNavigationWithinZoneAction(FocusZoneAction action
...
+{
+    if([view isKindOfClass:[RCTFocusZone class]])
+    {
+        // FocusZone has no elements
+        if ([[view subviews] count] < 1)
+        {
+            return YES;
+        }
+        
+        NSView* keyView = GetFirstKeyViewWithin(view);
```
Comment (harrieshin) [inline packages/components/FocusZone/macos/RCTFocusZone.m:171]: super nit: * location

File: packages/components/FocusZone/macos/RCTFocusZone.m [hunk 3]
```diff
@@ -157,10 +157,32 @@ static inline BOOL IsHorizontalNavigationWithinZoneAction(FocusZoneAction action
 	return nil;
 }
 
+/// Bypass FocusZone if it's empty or has no focusable elements
+static BOOL ShouldSkipFocusZone(NSView *view)
+{
+    if([view isKindOfClass:[RCTFocusZone class]])
+    {
+        // FocusZone has no elements
+        if ([[view subviews] count] < 1)
```
Comment (harrieshin) [inline packages/components/FocusZone/macos/RCTFocusZone.m:166]: this isn't combinable with line 171. if there isn't a FirstKeyviewwithin the view, there must be no elements in the focuszone?

File: packages/components/FocusZone/macos/RCTFocusZone.m (hunk 1 above)
Comment (chiuam) [reply packages/components/FocusZone/macos/RCTFocusZone.m:184]: I think this comment states the justification for overriding `acceptsFirstResponder`, though I think putting it above the function header would make more sense.

File: packages/components/FocusZone/macos/RCTFocusZone.m (hunk 3 above)
Comment (chiuam) [reply packages/components/FocusZone/macos/RCTFocusZone.m:166]: That's correct. My rationale behind adding that check was for readability and avoid calling `GetFirstKeyViewWithin` on an empty FocusZone.

**Review by Saadnajmi — APPROVED** [review]

File: packages/components/FocusZone/macos/RCTFocusZone.m (hunk 3 above)
Comment (Saadnajmi) [reply packages/components/FocusZone/macos/RCTFocusZone.m:166]: We could move the "check is empty logic" inside `GetFirstKeyViewWithin` too. But I think that function should be relatively cheap since it will do nothing if there are no subviews.

File: packages/components/FocusZone/macos/RCTFocusZone.m (hunk 1 above)
Comment (Saadnajmi) [reply packages/components/FocusZone/macos/RCTFocusZone.m:184]: The comment was added when Chris Pearson added support for `defaultTabbableElement` (before I think we just didn't override it). I also think it's worth updating the comment to clarify what cases we do/don't accept first responder.

File: packages/components/FocusZone/macos/RCTFocusZone.m (hunk 3 above)
Comment (chiuam) [reply packages/components/FocusZone/macos/RCTFocusZone.m]: I went with adding the "is empty" check inside `GetFirstKeyViewWithin`. Thought it'd benefit every caller that way

File: packages/components/FocusZone/macos/RCTFocusZone.m [hunk 4]
```diff
@@ -157,10 +162,26 @@ static inline BOOL IsHorizontalNavigationWithinZoneAction(FocusZoneAction action
 	return nil;
 }
 
+/// Bypass FocusZone if it's empty or has no focusable elements
+static BOOL ShouldSkipFocusZone(NSView *view)
+{
+    if([view isKindOfClass:[RCTFocusZone class]])
+    {
+        NSView *keyView = GetFirstKeyViewWithin(view);
+        // FocusZone is empty or has no focusable elements
```
Comment (harrieshin) [inline packages/components/FocusZone/macos/RCTFocusZone.m:171]: now this comment is redundant from the function's documentation because you hae improved the getfirstkeyviewwithin()

**Review by harrieshin — APPROVED** [review]

File: packages/components/FocusZone/macos/RCTFocusZone.m (hunk 4 above)
Comment (chiuam) [reply packages/components/FocusZone/macos/RCTFocusZone.m:171]: I think it's better to state the possibilities here so that we know all the cases why we want to reject first responder. Will prefer to leave it as is.
